lib/event-subscribers.loader.ts

```diff
@@ -30,27 +31,62 @@ export class EventSubscribersLoader
   loadEventListeners() {
     const providers = this.discoveryService.getProviders();
     const controllers = this.discoveryService.getControllers();
+
     [...providers, ...controllers]
-      .filter(wrapper => wrapper.isDependencyTreeStatic())
-      .filter(wrapper => wrapper.instance)
       .forEach((wrapper: InstanceWrapper) => {
-        const { instance } = wrapper;
-        const prototype = Object.getPrototypeOf(instance) || {};
-        this.metadataScanner.scanFromPrototype(
-          instance,
-          prototype,
-          (methodKey: string) =>
-            this.subscribeToEventIfListener(instance, methodKey),
-        );
+          if (wrapper.isDependencyTreeStatic() && wrapper.instance) {
+            this.subscribeSingleton(wrapper);
+          } else {
+            this.subscribeScoped(wrapper);
+          }
       });
+
+  }
+
+  private subscribeScoped(wrapper: InstanceWrapper) {
+    const prototype = wrapper.metatype.prototype;
+    if(!prototype)
+      return;
+    const instance = Object.create(prototype);
+
+    this.metadataScanner.scanFromPrototype(
+        instance,
+        prototype,
+         (methodKey: string) =>
+            this.subscribeToEventIfListener(instance, methodKey, async (...args: unknown[]) => {
+              if(args.length === 0)
+                return;
+              const event: any = args[0];
+              if (!(event instanceof ScopedEvent))
+                return;
+              const inst = await wrapper.host!.getProviderByKey<ModuleRef>(ModuleRef).instance.resolve(wrapper.metatype, event.context);
```

Comment:
This line is actually getting the request-scoped instance?

Reply:
Yep it is.
Here is breakdown:
`wrapper.host!.getProviderByKey<ModuleRef>(ModuleRef).instance` is just a lookup by key from dictionary.
ModuleRef is created and added as provider to Module (which is `wrapper.host`). So it is fast.

`resolve(wrapper.metatype, event.context)` will ask nestJs to lookup current request scope container for instance. If it is there - it is a simple dictionary lookup. If not, it will go into creation of instance like it does on each request.

Talking about performance. Typically you would use it to invert the dependency of changes!!! between different parts of code. 99% of cases it is not required to be high-performant. 
If you need high performance event emission is a no-go anyway, because you don't know about performance of handlers :) Separation of concerns as it is :)

Reply:
ok thank you for explaining. I was wondering about `getProviderByKey()` specifically, but good to hear that it's dictionary lookup